File: mapbox_maps_flutter_mobile/android/src/main/kotlin/com/mapbox/maps/mapbox_maps/MapProjectionController.kt
package com.mapbox.maps.mapbox_maps

import com.mapbox.geojson.Point
import com.mapbox.maps.MapboxMap
import com.mapbox.maps.mapbox_maps.pigeons.*

class MapProjectionController(private val mapboxMap: MapboxMap) : Projection {
  override fun getMetersPerPixelAtLatitude(latitude: Double, zoom: Double): Double {
    return mapboxMap.getMetersPerPixelAtLatitude(latitude, zoom)
  }

  override fun projectedMetersForCoordinate(coordinate: Point): ProjectedMeters {
    return mapboxMap.projectedMetersForCoordinate(coordinate).toFLTProjectedMeters()
  }

  override fun coordinateForProjectedMeters(projectedMeters: ProjectedMeters): Point {
    return mapboxMap.coordinateForProjectedMeters(projectedMeters.toProjectedMeters())
  }

  override fun unproject(
    coordinate: MercatorCoordinate,
    zoomScale: Double
  ): Point {
    return mapboxMap.unproject(coordinate.toMercatorCoordinate(), zoomScale)
  }

  override fun project(
    coordinate: Point,
    zoomScale: Double
  ): MercatorCoordinate {
    return mapboxMap.project(coordinate, zoomScale).toFLTMercatorCoordinate()
  }
}

File: mapbox_maps_flutter_mobile/ios/mapbox_maps_flutter_mobile/Sources/mapbox_maps_flutter_mobile/Classes/MapProjectionController.swift
import Foundation
import MapboxMaps
import UIKit
import Turf

class MapProjectionController: Projection {
    func getMetersPerPixelAtLatitude(latitude: Double, zoom: Double) throws -> Double {
        return MapboxMaps.Projection.metersPerPoint(for: latitude, zoom: zoom)
    }

    func projectedMetersForCoordinate(coordinate: Point) throws -> ProjectedMeters {
        let projectedMeters = MapboxMaps.Projection.projectedMeters(for: coordinate.coordinates)
        return ProjectedMeters(northing: projectedMeters.northing, easting: projectedMeters.easting)
    }

    func coordinateForProjectedMeters(projectedMeters: ProjectedMeters) throws -> Point {
        return Point(MapboxMaps.Projection.coordinate(for: projectedMeters.toProjectedMeters()))
    }

    func project(coordinate: Point, zoomScale: Double) throws -> MercatorCoordinate {
        let coordinate = MapboxMaps.Projection.project(coordinate.coordinates, zoomScale: zoomScale)
        return coordinate.toFLTMercatorCoordinate()
    }

    func unproject(coordinate: MercatorCoordinate, zoomScale: Double) throws -> Point {
        return Point(MapboxMaps.Projection.unproject(coordinate.toMercatorCoordinate(), zoomScale: zoomScale))
    }
}
